### DISCUSSION TITLE
Crop images with multiple and uneven borders

### DISCUSSION THREAD

**Discussion by ElvinJin** [ImageMagick/ImageMagick#7788] [Help]
### ImageMagick version

7.1.1-39

### Operating system, version and so on

macOS 15.1 

### Description

I have scans that contains 4 photos. Trying to split them into 4 images and remove the white background and non-even black frames (scanning holder). I tried deskew and crop, both works fine but couldn't figure out how to properly remove the black frames and just leave the polaroid in each image. Tried creating a mask but I think the uneven color of the black frame is not working with it. How I can achieve this?
 Attached the input image:
![img20241026_18525848](https://github.com/user-attachments/assets/8ceb423a-4d7b-4611-a4ff-43a7696ebe92)

First output after slicing:
![temp-0](https://github.com/user-attachments/assets/2a1d1483-a1b3-4ece-83d5-5b35b694bc7d)

Ideal output:
<img width="654" alt="Screenshot 2024-11-18 at 3 02 18 PM" src="https://github.com/user-attachments/assets/2ec05482-6bc4-4fc0-9248-8a1b41e69096">

Thank you.

**Comment by fmw42** [comment]
For the colors you have for your borders, this script seems to work for me on Mac OSX Ventura and IM 7.1.1.41

The first 3 lines can be chained into one command, if desired. The first command converts to 4 uniform colors (black, white, green, magenta). The next thresholds those colors so that dark goes to pure black and the rest goes to white. The next command trims your outer irregular white border. The next line gets the trim amount and stores it for later use. The next set of code uses connected components processing to find the crop coordinates (bounding boxes) of the 4 images with borders. The last "for" loop does a trim of the input and then crops to the bounding box coordinates to extract your 4 images. Note that I saved the images as JPG, since the PNG versions were to large.

```
magick img.png -kmeans 4 yy.png
magick yy.png -threshold 20% yyy.png
magick yyy.png -background white -define trim:percent-background=0% -trim +repage yyyy.png
trim=`magick yyy.png -background white -define trim:percent-background=0% -format "%@" info:`
echo $trim

bboxArr=(`magick yyyy.png \
-type bilevel \
-define connected-components:verbose=true \
-define connected-components:keep-top=4 \
-define connected-components:exclude-header=true \
-connected-components 4 null: | grep "gray(255)" | awk '{print $2}'`)
num=${#bboxArr[*]}
echo $num

for ((i=0; i<num; i++)); do
bbox=${bboxArr[$i]}
echo $bbox
magick img.png -crop $trim +repage -crop $bbox +repage yyyy-$i.jpg
done

```


![yyyy-0](https://github.com/user-attachments/assets/fc73875e-f6dc-4aa0-9e8c-4889a3d54a89)

![yyyy-1](https://github.com/user-attachments/assets/5e1dba2f-ce78-4249-90f9-f8a24476c944)

![yyyy-2](https://github.com/user-attachments/assets/37408af7-4c0e-41ea-ae75-c5115350b2c8)

![yyyy-3](https://github.com/user-attachments/assets/ca6b4e14-d1dd-46de-848e-08a9e8a42432)
